src/rp_data_tree.c

```diff
@@ -0,0 +1,476 @@
+/**
+ * @file rp_data_tree.c
+ * @author Rastislav Szabo <raszabo@cisco.com>, Lukas Macko <lmacko@cisco.com>
+ * @brief 
+ *
+ * @copyright
+ * Copyright 2015 Cisco Systems, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "rp_data_tree.h"
+#include "xpath_processor.h"
+#include "sr_common.h"
+
+
+/**
+ * @brief Creates part of xpath for leaf and container nodes. Arguments specify if the namespace and trailing slash
+ * should be included.
+ */
+static int
+rp_dt_create_xpath_for_cont_leaf_node(const struct lyd_node *data_tree, char **xpath, bool namespace, bool trailing_slash)
+{
+    CHECK_NULL_ARG2(data_tree, xpath);
+    char *s = NULL;
+    size_t len = 1; /* terminating null byte*/
+    size_t ns_len = 0;
+    size_t node_len = 0;
+
+    /* calculate length */
+    if (namespace){
+        ns_len = strlen(data_tree->schema->module->name) + 1; /*namespace + colon*/
```

Comment:
NULL check data_tree->schema, data_tree->schema->module and data_tree->schema->module->name before this strlen. You can use  CHECK_NULL_ARG3 macro.